Version 2.1.0 Class Maintenance and Performance Enhancements

GuillaumeCleme opened the pull request:
Implemented PhantomJSConstants class, and performance enhancements. Setup now checks existence of binaries before extraction, and runs using a single executable instead of a transaction based extraction. Multi-threadness is handled via OS using single executable. Packages have also been created for class segmentation and maintenance.

Additional load test has been added to `TestPhantomJS.class` to test multiple executable invocations using a single binary file.

GuillaumeCleme committed:
Performing class maintenance and creating packages

GuillaumeCleme committed:
Refactor methods into setup classes and clean up imports.

GuillaumeCleme committed:
Implemented PhantomJSConstants class, and performance enhancements. Setup now checks existence of binaries before extraction, and runs using a single executable instead of a transaction based extraction. Multi-threadness is handled via OS using single executable.

GuillaumeCleme committed:
Merge branch 'feature/Version2.1.0' into develop

GuillaumeCleme committed:
Updated TestPhantomJS.java test

Updated TestPhantomJS.java testFooterAndHeader() assertion for Travis build.

moodysalem commented:
Thanks for the work, I'll look at it later tonight

moodysalem commented on src/main/java/com/moodysalem/phantomjs/wrapper/CommandLineArgument.java:
```diff
@@ -25,6 +25,10 @@ public CommandLineArgument(String template, String key, Object val) {
         return objMap;
     }
+    
+    public static String wrapTemplateName(String templateName){
```
@GuillaumeC91 Bit of a confusing method name-maybe wrapCommandLineArgumentName or something

GuillaumeC91 replied:
@moodysalem Renamed method to `wrapCommandLineArgumentName`

moodysalem commented on src/main/java/com/moodysalem/phantomjs/wrapper/beans/PhantomJSExecutionResponse.java:
```diff
@@ -0,0 +1,34 @@
+	private String stdErr;
+	
+	public PhantomJSExecutionResponse(int exitCode, String stdOut, String stdErr) {
+		super();
```
no need for super

GuillaumeC91 replied:
@moodysalem Removed, came from automated generation.

moodysalem commented on src/main/java/com/moodysalem/phantomjs/wrapper/PhantomJS.java:
```diff
@@ -1,405 +1,286 @@
+	 * LoggerOutputStream private nested class override
+	 *
+	 */
+	private static class LoggerOutputStream extends LogOutputStream {
```
Nice enhancement

moodysalem commented on src/test/java/TestPhantomJS.java:
```diff
@@ -65,7 +89,60 @@ public void testFooterAndHeader() throws IOException, RenderException {
             PDDocument doc = PDDocument.load(is);
-            assertTrue(doc.getNumberOfPages() == 2);
+            doc.close();
+            assertTrue(doc.getNumberOfPages() == 2); //Only 1 page is generated on Windows. This seems platform dependent.
```
If you look at render.js, we manipulate the zoom level of the document depending on the OS in order to get consistent results across platforms

GuillaumeC91 replied:
@moodysalem Left the tested value to 2 in order to satisfy Travis CI tests. Will look into possibly tweaking this setting to get consistent results on Windows.

moodysalem commented on src/main/java/com/moodysalem/phantomjs/wrapper/PhantomJSSetup.java:
```diff
@@ -0,0 +1,173 @@
+	protected final static Logger logger = Logger.getLogger(PhantomJSSetup.class.getName());
+	
+	// this will store a reference to the executable phantomjs binary after we unzip the resource
+    protected static File PHANTOM_JS_BINARY = null;
```
We should make this final and initialize it in a static block for this class-versus initializing it in the PhantomJS class

GuillaumeC91 replied:
@moodysalem Made class a _public final_ and refactored the static initialization block to `PhantomJSSetup`.

moodysalem commented on src/main/java/com/moodysalem/phantomjs/wrapper/helpers/StreamHelpers.java:
```diff
@@ -0,0 +1,26 @@
+import java.io.IOException;
+import java.io.InputStream;
+
+public class StreamHelpers {
```
abstract this

GuillaumeC91 replied:
@moodysalem Done, currently not in use but it was refactored from the main PhantomJS class.

moodysalem commented:
@GuillaumeC91 
One thing I'm not sure about is removing the timestamp from the path

I wanted to make sure each jvm was working in its own separate tmp directory, otherwise it's possible that an input html file or header and footer file could be overwritten when two separate JVMs call render with the same atomic long value

GuillaumeCleme commented:
@moodysalem 
Interesting observation, I did not really factor in multiple JVMs. I had in mind being able to support multiple binary invocations without extraction every time. If multiple JVMs sounds like a plausible use case, we may have to add that back in. Perhaps something can be done using a [VMID](http://docs.oracle.com/javase/6/docs/api/java/rmi/dgc/VMID.html)

I will take a look at the other enhancements this evening and commit minor fixes.

Cheers,

moodysalem commented:
To clarify, with the current implementation, the extraction should only happen once when the PhantomJS class is initialized since that code all occurs in static blocks-so multiple invocations of render should only ever result in one extraction (unless the JVM reinitializes the class for some reason)

It would be OK IMO to reuse the phantomjs binaries across JVMs and just establish different tmp directories for the temporary html and footer/header files (preferably using UUID::randomUUID rather than timestamps)

GuillaumeCleme commented:
Added a final UUID as part of constants that get's used during the creation of the script, source, and output folder. This includes and new shutdown hook in charge of cleaning of the JVM specific folder at JVM shutdown.

Commit:
Support for multiple JVM invocation using same binaries. Added a final UUID as part of constants that get's used during the creation of the script, source, and output folder. This includes and new shutdown hook in charge of cleaning of the JVM specific folder at JVM shutdown. Minor abstraction changes and cleanup.

moodysalem commented on src/main/java/com/moodysalem/phantomjs/wrapper/PhantomJSSetup.java:
```diff
@@ -0,0 +1,215 @@
+	protected final static Logger logger = Logger.getLogger(PhantomJSSetup.class.getName());
+
+	// this will store a reference to the executable phantomjs binary after we unzip the resource
+	protected static File PHANTOM_JS_BINARY = null;
```
It should be possible to make this PHANTOM_JS_BINARY field final and initialize it in the static block-just so we don't have to worry about it ever being modified after class initialization

GuillaumeC91 replied:
@moodysalem Gotcha. I've made the whole class final, but maybe being able to support re-initialization is something we need. As for the `PHANTOM_JS_BINARY` being final, we would have to initialize it by making a `private function` call correct? Something like so:

`protected static File PHANTOM_JS_BINARY = initialize();`

moodysalem replied:
Yeah that should work-not as worried about the class being final since it's less likely IMO that someone would accidentally subclass it and cause issues vs. accidentally assigning the variable a new value

GuillaumeCleme committed:
PhantomJS Binary Initialization

New initialization private method with final file field. Initialization may now throw exception if initialization fails.

GuillaumeCleme committed:
PhantomJSSetup fixups to static final binaries initiation.

GuillaumeCleme commented:
Mobile/Plain text code editor [fail](https://github.com/moodysalem/java-phantomjs-wrapper/pull/1/commits/ab4e01a852113e8fa849480173db878d19091f48) :-1:

GuillaumeCleme commented:
Hey @moodysalem. Anything else you'd like to see implemented?

moodysalem commented:
@GuillaumeC91 I merged in the work and made some additional changes today, then deployed it to central in 2.1.3

GuillaumeCleme commented:
@moodysalem Awesome. Looks good! One tiny little detail; forgot to update the ReadMe.md with the new return type of `exec()` from `int` to `PhantomJSExecutionResponse`. No big deal I suppose.
